### ISSUE TITLE
Support pointfree notation with curried assertions?

### ISSUE THREAD

**Issue by Siilwyn** [avajs/ava#1991]
Would you consider currying ava assertions putting the data argument last with an additional export?

## Usage example
```js
const test = require('ava/curried');

test('something', t => Promise.resolve('hi').then(t.is('hi'));
test('something', t => Promise.resolve('hi').then(t.is('hi', 'this is a message'));
```

Instead of:
```js
const test = require('ava');

test('something', t => Promise.resolve('hi').then(result => t.is(result, 'hi'));
test('something', t => Promise.resolve('hi').then(result => t.is(result, 'hi', 'this is a message'));
```

## Implementation outline
```js
const is = (expected, [message]) => (value) => {};
```

**Comment by foxbunny** [comment]
It probably won't be as simple as what you describe as it would need to support both binary and unary forms. It'll probably end up being not worth it. Partially applying `t.is` is probably simplest here:

```
test('something', t => Promise.resolve('hi').then(t.is.bind(t, 'hi'));

**Comment by Siilwyn** [comment]
@foxbunny could you expand on what you mean with support for binary and unary? To clarify I wouldn't want variadic currying, by exporting it separately this won't be needed.

So never do this:
```js
const test = require('ava/curried');
test('something', t => Promise.resolve('hi').then(result => t.is('hi', 'this is a message', result));
```

Because using the default export already offers a clean way to do this.

**Comment by foxbunny** [comment]
Oh, ok, I get what you mean. Still, there's an increased maintenance cost doing that where partial application is already available natively and can solve the issue. Of course, I'm not the maintainer, so it's not my problem. Just pointing it out.

**Comment by Siilwyn** [comment]
Yes thanks for the input! Maintenance cost is an understandable concern, in this case it means when changing or adding assertion (arguments) it would also need to be changed in the 'curry wrapper'. Though since ava is already quite mature I don't see this happen often.

Binding every assertion would be a solution but I'd rather not since it still adds 'noise' as much as passing the argument with an arrow function.

**Comment by foxbunny** [comment]
Personally, I'd rather that ava developers have an easier time maintaining the library because that means it will be more stable for me. I'll trade stable for ergonomics every time, especially since this is a test framework -- it's supposed to be fool-proof. Why not just start a 3rd party project for the curried version?

**Comment by Siilwyn** [comment]
Having a 3rd party project would generate even more work, and as I said before this won't add much maintenance time. I understand that this adds complexity like any new feature, and like any new feature the authors can make the call if it is worth adding. It would be very nice to write tests in a functional style. I don't see how this affects the stability of ava.

**Comment by novemberborn** [comment]
I can see the utility, but I'm not sure about the actual use cases. @Siilwyn in your example you have promise code, which is better served with `async` / `await`. Assertions also take an optional final assertion message, which makes passing them to array functions impossible (since a number will be passed to the message, which if we're not already throwing on we *should* throw on).

**Comment by Siilwyn** [comment]
Ah okay, bummer but understandable. :)
